### PR TITLE
Do not suggest "is a function" for free variables

### PR DESCRIPTION
**Pull request by notriddle**
Part of #82323

### CODE REVIEW

File: src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// check-fail
```
Comment (petrochenkov) [inline src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs:1]: `// check-fail` annotations are unnecessary, it's the default for ui tests.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 2]
```diff
@@ -40,6 +40,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                // type resolution always gives a "maybe" here.
+                if self
+                    .autoderef(span, ty)
+                    .any(|(ty, _)| matches!(ty.kind(), ty::Error(_) | ty::Infer(_)))
+                {
+                    return false;
+                }
+
                 self.autoderef(span, ty).any(|(ty, _)| {
                     self.probe(|_| {
```
Comment (petrochenkov) [inline compiler/rustc_typeck/src/check/method/suggest.rs:55]:
```suggestion
                    !matches!(ty.kind(), ty::Error(_) | ty::Infer(_)) && self.probe(|_| {
```
It looks like there's no need to walk the derefs twice.

Comment (notriddle) [reply compiler/rustc_typeck/src/check/method/suggest.rs:55]: This doesn't work for the `Vec` test case.

<details>

```
error[E0412]: cannot find type `T` in this scope
 --> src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs:2:13
  |
2 | impl Struct<T>
  |     -       ^ not found in this scope
  |     |
  |     help: you might be missing a type parameter: `<T>`

error[E0412]: cannot find type `T` in this scope
 --> src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs:7:5
  |
7 |     T: Copy,
  |     ^ not found in this scope

error[E0282]: type annotations needed
  --> src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs:14:31
   |
14 |     fn method(v: Vec<u8>) { v.len(); }
   |                               ^^^ cannot infer type
   |
   = note: type must be known at this point

 INFO rustc_typeck::check::method::suggest check deref std::vec::Vec<u8> impl FnOnce
error[E0599]: no method named `len` found for struct `Vec<u8>` in the current scope
  --> src/test/ui/functions-closures/fn-help-with-err-generic-is-not-function.rs:14:31
   |
14 |     fn method(v: Vec<u8>) { v.len(); }
   |                               ^^^ private field, not a method
   |
   = note: `v` is a function, perhaps you wish to call it

┐rustc_typeck::check::coercion::coerce a=(), b=()
┘
```

</details>

It looks like, based on the added trace I put in here, the `Vec<u8>` itself returns true from `predicate_may_hold`. It derefs to an error type, so if we check the deref chain completely we can avoid this bug, but by doing it this way, we cause it to short-circuit at the first matching type.

Comment (petrochenkov) [reply compiler/rustc_typeck/src/check/method/suggest.rs:56]: Ok.

**Comment by petrochenkov** [comment]
r=me after squashing commits.

**Pull request merged**